src/adc.rs

```diff
@@ -0,0 +1,407 @@
+                        7 => self
+                            .rb
+                            .smpr2
+                            .modify(|_, w| unsafe { w.smp7().bits(self.sample_time.into()) }),
+                        8 => self
+                            .rb
+                            .smpr2
+                            .modify(|_, w| unsafe { w.smp8().bits(self.sample_time.into()) }),
+                        9 => self
+                            .rb
+                            .smpr2
+                            .modify(|_, w| unsafe { w.smp9().bits(self.sample_time.into()) }),
+
+                        10 => self
+                            .rb
+                            .smpr1
+                            .modify(|_, w| unsafe { w.smp10().bits(self.sample_time.into()) }),
+                        11 => self
+                            .rb
+                            .smpr1
+                            .modify(|_, w| unsafe { w.smp11().bits(self.sample_time.into()) }),
+                        12 => self
+                            .rb
+                            .smpr1
+                            .modify(|_, w| unsafe { w.smp12().bits(self.sample_time.into()) }),
+                        13 => self
+                            .rb
+                            .smpr1
+                            .modify(|_, w| unsafe { w.smp13().bits(self.sample_time.into()) }),
+                        14 => self
+                            .rb
+                            .smpr1
+                            .modify(|_, w| unsafe { w.smp14().bits(self.sample_time.into()) }),
+                        15 => self
+                            .rb
+                            .smpr1
+                            .modify(|_, w| unsafe { w.smp15().bits(self.sample_time.into()) }),
+
+                        _ => unreachable!(),
+                    }
+
+                    return;
+                }
+
+                fn convert(&mut self, chan: u8) -> u16 {
+                    self.rb.cr2.modify(|_, w| w.align().bit(self.align.into()));
+                    self.set_chan_smps(chan);
+                    self.rb.sqr3.modify(|_, w| unsafe { w.sq1().bits(chan) });
+
+                    // ADC start conversion of regular sequence
+                    self.rb.cr2.modify(|_, w| w.swstart().set_bit());
+                    while self.rb.cr2.read().swstart().bit_is_set() {}
+                    // ADC wait for conversion results
+                    while self.rb.sr.read().eoc().bit_is_clear() {}
+
+                    let res = self.rb.dr.read().data().bits();
+                    res
+                }
+            }
+
+            impl<WORD, PIN> OneShot<$ADC, WORD, PIN> for Adc<$ADC>
+            where
+                WORD: From<u16>,
+                PIN: Channel<$ADC, ID = u8>,
+                {
+                    type Error = ();
+
+                    fn read(&mut self, _pin: &mut PIN) -> nb::Result<WORD, Self::Error> {
```

Comment:
If I am reading this correctly, this function is blocking despite returning an `nb::Result` which seems weird to me.

Reply:
Well, this is how _read_ is declared in embedded-hal OneShot trait. This nb::Result is not necessarily about blocking or non-blocking function. This is just a minimal extension of std::io::Result which adds WouldBlock error code, see [nb docs](https://docs.rs/nb/0.1.1/nb/).

Reply:
Maybe I'm misinterpreting things then, but I as a user would assume that a function returning `nb::Result` is actually non-blocking.

I could see someone implementing a device driver using the `embedded_hal` traits that relies on the ADCs being non blocking for doing something like concurrent reads.

Reply:
Hi @TheZoq2, @therealprof , and all,
Ok, lets finally decide on this one. To block, or not to block: that is the question.
* To block until conversion is completed
Apparently nb::Result can be returned from blocking call: in this case we do not return WouldBlock. So far this approach has been taken by other stm32 HALs including stm32f0xx-hal, stm32f4xx-hal, stm32l0xx-hal
* To use completion bits and return WouldBlock if conversion is not ready
This approach has certain advantages, see @TheZoq2 comments. In this case I will adapt the code from #19 and update adc example.

Comments ? Thoughts ?

Regards,
Sergey

Reply:
@geomatsi If you can implement it truly non-blocking, then go for it. I believe the blocking in the other hals code only happens because any other implementation is not race free. The problem here usually is to know whether the call is supposed to start a new sampling or collect the data of a previous sampling... Of course we can always start blocking and improve in later iteration if someone can think of a clever solution.

Reply:
I agree, non blocking would be ideal but it does have some pitfalls, for example [this one](https://github.com/japaric/stm32f103xx-hal/pull/114#issuecomment-450547784) from my original PR.

If you can find a nice way to make it non-blocking, I would prefer it, but i'm also fine with the current solution